### ISSUE TITLE
Resizing to network disk

### ISSUE THREAD

**Issue by richdunajewski** [lovell/sharp#1551]
I'm running sharp under a lightweight Debian build, but the same situation occurs in a Docker environment.

When writing to a network drive (NFS mount), I see that sharp and/or libvips is holding open a bunch of files. Initially suspecting performance issues with the NFS server, we've eliminated that as a possibility. Sharp goes to write the file, and the operation will hang. The only way to free up the hung operation is to force unmount the share. 

The resize operations occur in a loop, but are limited to 4 concurrent operations at a time to prevent overwhelming the system. Once we switch over to a local directory, all issues disappear.

The suspicion is that sharp may be returning a successful callback result before the NFS synchronizes and due to the loop, eventually the system is overloaded and hangs. 

How does vips and sharp handle such a scenario?

**Comment by lovell** [comment]
Hello, have you tried preventing caching within libvips via `sharp.cache(false)`?

https://sharp.pixelplumbing.com/en/stable/api-utility/#cache

**Comment by richdunajewski** [comment]
Yes, we noticed that files seem to get left open if we don't disable the cache. So we disable it.

Is there any harm in calling `sharp.cache(false)` multiple times? I believe we called it after each resize operation to "flush" the cache, but maybe that's torquing around the vips options repeatedly?

**Comment by richdunajewski** [comment]
Could the concurrency option be compounding the issue? We have a worker process for each of the 4 operations, and so sharp is being instantiated 4 times. Could it be trying to spawn up to x operations based on the number of CPU cores? Sounds like we'd want a concurrency of 1 per worker to make sure really only 4 operations are kicked off at a time.

**Comment by lovell** [comment]
There's no harm calling `sharp.cache(false)` repeatedly.

If "concurrency" in this context refers to `sharp.concurrency()` then that relates to the number of threads that libvips should spawn for an image pipeline, which is every time your code calls `toFile()`.

https://sharp.pixelplumbing.com/en/stable/api-utility/#concurrency

You're unlikely to see much benefit by spawning multiple worker processes.

**Comment by jcupitt** [comment]
Hello, 

Something like imagemagick will do a series of reads to pull the image in, then a series of writes to send the result back. libvips will issue a similar set of read and write calls, but will overlap them. Could that make a difference? I'm not sure. 

NFS is asynchronous, of course, so write returns as soon as your request has been accepted, there is no wait for the network to actually process it. If you are calling write faster than the network can handle it, you can get the output buffers building up. There are some NFS options to control this process, you'd need to ask an experienced admin.  

If you are also reading from NFS, you'll probably find you are fetching the source image each time you make a smaller derivative. You might get better performance by reading the source image to a memory area, then shrinking from that to the output.

**Comment by jcupitt** [comment]
Oh, just saw the bit about caching, sorry. 

I wouldn't disable caching. Instead, fetch the NFS source once to a memory area and then start a set of workers, one for each derivative.

Your network is very likely to be your bottleneck, so optimisation and parallelisation probably won't help much.

**Comment by richdunajewski** [comment]
We found the culprit. After some back and forth with the team who provides the NFS server, we discovered that there is a bug affecting NFS v4.1 that we hit. Something about a client query that would fail, and cause the client to go into a loop of sending further requests. We forced our connection to 4.0 and the issue disappeared. 

So I'm closing this issue as it wasn't sharp causing the problem. Thanks for the suggestions, it helped us narrow down the problem.
